protocols/kad/src/query/peers/closest.rs

```diff
@@ -0,0 +1,694 @@
+            progress = self.closest_peers.keys().next() == Some(&distance)
+                || num_closest < self.config.num_results;
+        }
+
+        // Update the iterator state.
+        self.state = match self.state {
+            State::Iterating { no_progress } => {
+                let no_progress = if progress { 0 } else { no_progress + 1 };
+                if no_progress >= self.config.parallelism {
+                    State::Stalled
+                } else {
+                    State::Iterating { no_progress }
+                }
+            }
+            State::Stalled =>
+                if progress {
+                    State::Iterating { no_progress: 0 }
+                } else {
+                    State::Stalled
+                }
+            State::Finished => State::Finished
+        }
+    }
+
+    /// Callback for informing the iterator about a failed request to a peer
+    /// that the iterator is waiting on.
+    ///
+    /// After calling this function, `next` should eventually be called again
+    /// to advance the state of the iterator.
+    ///
+    /// If the iterator is finished, it is not currently waiting for a
+    /// result from `peer`, or a result for `peer` has already been reported,
+    /// calling this function has no effect.
+    pub fn on_failure(&mut self, peer: &PeerId) {
+        if let State::Finished = self.state {
+            return
+        }
+
+        let key = Key::from(peer.clone());
+        let distance = key.distance(&self.target);
+
+        match self.closest_peers.entry(distance) {
+            Entry::Vacant(_) => return,
+            Entry::Occupied(mut e) => match e.get().state {
+                PeerState::Waiting(_) => {
+                    debug_assert!(self.num_waiting > 0);
+                    self.num_waiting -= 1;
+                    e.get_mut().state = PeerState::Failed
+                }
+                PeerState::Unresponsive => {
+                    e.get_mut().state = PeerState::Failed
+                }
+                _ => {}
+            }
+        }
+    }
+
+    /// Returns the list of peers for which the iterator is currently waiting
+    /// for results.
+    pub fn waiting(&self) -> impl Iterator<Item = &PeerId> {
+        self.closest_peers.values().filter_map(|peer|
+            match peer.state {
+                PeerState::Waiting(..) => Some(peer.key.preimage()),
+                _ => None
+            })
+    }
+
+    /// Returns the number of peers for which the iterator is currently
+    /// waiting for results.
+    pub fn num_waiting(&self) -> usize {
+        self.num_waiting
+    }
+
+    /// Returns true if the iterator is waiting for a response from the given peer.
+    pub fn is_waiting(&self, peer: &PeerId) -> bool {
+        self.waiting().any(|p| peer == p)
+    }
+
+    /// Advances the state of the iterator, potentially getting a new peer to contact.
+    pub fn next(&mut self, now: Instant) -> PeersIterState {
```

Comment:
How is the `peer_timeout` enforced? What tells the caller that they should call `next()` after a certain number of seconds have passed?

Reply:
This module is agnostic to how time progresses (and at what granularity), that is entirely left up the code using it (which makes for side-effect-free code and easy testing). Note that these modules are private to the crate and this module in particular is really just the previous `query.rs` with some renaming, i.e. this isn't really new code.

Reply:
> that is entirely left up the code using it

Which raises the question: how does the code using it know when the timeout has expired and that we should poll the `ClosestPeersIter` again?

Reply:
If I understand the question correctly, it does not need to know / does not care. The code using it only cares about driving a query to completion by sending requests to all peers yielded by the query and reporting the results back. Concretely, `Kademlia::poll()` polls the `QueryPool` to obtain peers to send requests to. The QueryPool in turn looks through its queries, advancing the peer "iterators" to get the next peers to send requests to. Whenever in an execution of `Kademlia:poll` all queries in the pool yield no more peers (e.g. parallelism is at capacity), the pool itself yields `AtCapacity` or `Waiting(None)` and `Kademlia::poll` may return `NotReady` at this point. There is I/O in progress, all the necessary task wakeup is driven by the I/O. The QueryPool and the queries inside with their peer selection strategies just implement the state machines for advancing the queries. `Kademlia::poll` passes down the current `Instant` whenever it asks the `QueryPool` for new stuff to do and in the case of the `ClosestPeersIter`, this influences when the iterator terminates, as peers for whom a result was not delivered back within the `peer_timeout` are merely not being waited for when evaluating the termination conditions (but delayed results are still accepted, if they come in before the query is done).

I hope that makes sense.

Reply:
That doesn't answer my question that I had in mind :sweat_smile: but I'm going to trust you on this one.